u_parrots.go

```diff
@@ -938,3 +941,367 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+						return false
+					}
+					m.supportedVersions = append(m.supportedVersions, vers)
+				}
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &SupportedVersionsExtension{m.supportedVersions})
+				// end utls specific
+
+			case extensionCookie:
+				// RFC 8446, Section 4.2.2
+				if !readUint16LengthPrefixed(&extData, &m.cookie) ||
+					len(m.cookie) == 0 {
+					return false
+				}
+
+				// utls specific
+				// TODO: I believe this should never be sent with initial client hello
+				// end utls specific
+
+			case extensionKeyShare:
+				// RFC 8446, Section 4.2.8
+				var clientShares cryptobyte.String
+				if !extData.ReadUint16LengthPrefixed(&clientShares) {
+					return false
+				}
+				for !clientShares.Empty() {
+					var ks keyShare
+					if !clientShares.ReadUint16((*uint16)(&ks.group)) ||
+						!readUint16LengthPrefixed(&clientShares, &ks.data) ||
+						len(ks.data) == 0 {
+						return false
+					}
+					m.keyShares = append(m.keyShares, ks)
+				}
+
+				// utls specific
+				var ks keyShares = m.keyShares
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &KeyShareExtension{ks.ToPublic()})
+				// end utls specific
+
+			case extensionEarlyData:
+				// RFC 8446, Section 4.2.10
+				m.earlyData = true
+
+				// utls specific
+				// TODO: read this into a generic extension
+				// end utls specifis
+
+			case extensionPSKModes:
+				// RFC 8446, Section 4.2.9
+				if !readUint8LengthPrefixed(&extData, &m.pskModes) {
+					return false
+				}
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &PSKKeyExchangeModesExtension{m.pskModes})
+				// end utls specific
+
+			case extensionPreSharedKey:
+				// RFC 8446, Section 4.2.11
+				if !extensions.Empty() {
+					return false // pre_shared_key must be the last extension
+				}
+				var identities cryptobyte.String
+				if !extData.ReadUint16LengthPrefixed(&identities) || identities.Empty() {
+					return false
+				}
+				for !identities.Empty() {
+					var psk pskIdentity
+					if !readUint16LengthPrefixed(&identities, &psk.label) ||
+						!identities.ReadUint32(&psk.obfuscatedTicketAge) ||
+						len(psk.label) == 0 {
+						return false
+					}
+					m.pskIdentities = append(m.pskIdentities, psk)
+				}
+				var binders cryptobyte.String
+				if !extData.ReadUint16LengthPrefixed(&binders) || binders.Empty() {
+					return false
+				}
+				for !binders.Empty() {
+					var binder []byte
+					if !readUint8LengthPrefixed(&binders, &binder) ||
+						len(binder) == 0 {
+						return false
+					}
+					m.pskBinders = append(m.pskBinders, binder)
+				}
+
+				// utls specific
+				// TODO: do we want to read this into a generic extension or should we leave it alone
+				// end utls specifis
+
+			default:
+				// Ignore unknown extensions.
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &GenericExtension{extension, extData})
```

Comment:
Yeah so at the very least, I think we'd want to carry through GREASE extensions (they're currently being handled by the `GenericExtension` approach).

It's a little hard to think abstractly about what kinds of other generic extensions might be and whether we'd want to copy/record them. Two thoughts on that front:
- Maybe we should try and do some larger data gathering on extensions that aren't parsed here?
- Maybe we should somewhat frequently send active queries and record client hellos so that we can update the spec as we see new hellos (so that changing hellos are accurately reflected??)

Reply:
I think a look through the extensions in the spec is worthwhile.  Also although we don't know the exact distribution or what might show up tomorrow, we can/should certainly test with the set of highly likely browsers especially in china ... chrome, uc browser, qq, 360 safe, sougou, safari, ie, edge, firefox, edge ... mobile variants.

Reply:
I'm not sure carrying through fixed grease values is the right idea if the hello is used as a template for multiple requests...  That's another sorta gray area where there appears to be /a/ behavior available (what BoringSSL does... random legal value for each new hello (with some caveats)) though it /could/ vary in others.

Reply:
It looks like your other stuff supports re-greasing, but I'm not sure this one does... I think assigning a new grease value depends specifically seeing a UTLSGreaseExtension in this case? https://github.com/getlantern/utls/blob/master/u_parrots.go#L627

Reply:
> Maybe we should try and do some larger data gathering on extensions that aren't parsed here?

> Maybe we should somewhat frequently send active queries and record client hellos so that we can update the spec as we see new hellos (so that changing hellos are accurately reflected??)

Definitely agree on both points.  I'll likely be testing lots of possible OS/browser combinations anyway, so that should give us a good sample to play with.  I think it would also make a lot of sense for the client to report any hellos it is unable to parse.  In that case, we could just have some reasonable defaults for the client to fallback to (these could be built-in, distributed by the config server, or both).  The client could send the hello along with browser and OS information.  We could collect these to study and figure out what we're missing.

I suspect that by testing a good range of browser and OS combinations ourselves pre-deployment, we will find most of the missing extensions.  The reports from the clients will probably help us clean up the majority of the rest, leaving us with a few rare edge cases.  Like you say, we also want to track any changes which lead to parsing errors and the reports from the clients will help us with that as well.

Reply:
And I think all of that together means that it's unlikely we'd ever want to blindly carry something through.

Reply:
> 

It looks like your other stuff supports re-greasing, but I'm not sure the extensions re-grease when using a GenericExtension... I think assigning a new grease value depends specifically seeing a UTLSGreaseExtension in this case? https://github.com/getlantern/utls/blob/master/u_parrots.go#L627

Good point - I can just check if the extension id is a grease value and make it a grease extension instead. Thanks for the note on that!

Reply:
> And I think all of that together means that it's unlikely we'd ever want to blindly carry something through.

I _thiiink_ there might be a couple other (non-GREASE, non-padding) places where I'm currently leaning on the `GenericExtension` in order to properly form the hello. Let me add some logging back in so I can double check that

Reply:
@forkner I've updated the code to properly handle the `UtlsGreaseExtension` instead of shoving it into a `GenericExtension` which I think will take care of regreasing for us. Thanks for that tip!